### PR TITLE
Refactor Query Tree visitor

### PR DESCRIPTION
**Pull request by novikd**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Review by antonio2368 — COMMENTED** [review]
LGTM in general

are we planning to remove all other visitors?
We have so many of them...

File: src/Analyzer/InDepthQueryTreeVisitor.h [hunk 1]
```diff
@@ -235,4 +236,81 @@ class InDepthQueryTreeConditionalVisitor : public InDepthQueryTreeVisitor<InDept
...
+    explicit QueryTreeVisitor(ContextPtr context_)
+        : current_context(std::move(context_))
+    {}
+
+    bool needApply(QueryTreeNodePtr & node)
+    {
+        return getImpl().needApply(node);
+    }
+
+    bool shouldSkipSubtree(QueryTreeNodePtr & parent, size_t subtree_index)
```
[the remark this answers is not in the document]
Comment (novikd) [reply src/Analyzer/InDepthQueryTreeVisitor.h:252]: Do you have some test in mind?

[the remark this answers is not in the document]
Comment (antonio2368) [reply src/Analyzer/InDepthQueryTreeVisitor.h:252]:
```
select * from view(select * from system.numbers where number = 2 or number = 3)

DB::Exception: Method getResultType is not supported for IDENTIFIER query node. (UNSUPPORTED_METHOD)
```
or
```
select * from view(select * from system.numbers where number = 2 and number = 3)

DB::Exception: Function node with name 'and' is not resolved. (UNSUPPORTED_METHOD)
```

**Comment by Algunenano** [comment]
What's the status of this? It would help close https://github.com/ClickHouse/ClickHouse/issues/49773

**Pull request merged**